horovod/common/controller.h

```diff
@@ -0,0 +1,155 @@
+
+  virtual void CrossRankBitwiseOr(std::vector<long long>& bitvector,
+                                  int count) = 0;
+
+  virtual void SynchronizeParameters() = 0;
+
+  // Concrete controller functions
+  ResponseList ComputeResponseList(std::atomic_bool& shut_down);
+
+  int64_t TensorFusionThresholdBytes();
+
+  void SetRank(const int* ranks, int nrank) {
+    for (auto i = 0; i < nrank; ++i) {
+      ranks_.push_back(ranks[i]);
+    }
+  };
+
+  void SetTimelineEnabled(bool value) { timeline_enabled_ = value; }
+  int GetLocalSizeAtCrossRank(int i);
+  std::vector<int>& GetRanks() { return ranks_; };
+  int GetRank() { return rank_; };
+  int GetLocalRank() { return local_rank_; };
+  int GetCrossRank() { return cross_rank_; };
+  int GetSize() { return size_; };
+  int GetLocalSize() { return local_size_; };
+  int GetCrossSize() { return cross_size_; };
+  const std::vector<int>& GetLocalCommRanks() { return local_comm_ranks_; };
+  bool IsCoordinator() const { return is_coordinator_; };
+  bool IsHomogeneous() const { return is_homogeneous_; };
+
+  StallInspector& GetStallInspector() { return stall_inspector_; };
+
+protected:
+  virtual void RecvReadyTensors(std::vector<std::string>& ready_to_reduce,
+                                std::vector<RequestList>& ready_list) = 0;
+
+  virtual void SendFinalTensors(ResponseList& response_list) = 0;
+
+  virtual void SendReadyTensors(RequestList& message_list) = 0;
+
+  virtual void RecvFinalTensors(ResponseList& response_list) = 0;
+
+  Response ConstructResponse(std::string& name);
+
+  void CoordinateCacheAndState(CacheCoordinator& cache_coordinator);
+
+  ResponseList FuseResponses(std::deque<Response>& responses);
+
+  int64_t
+  TotalByteSizeOfAllgatherOutput(const std::vector<int64_t>& tensor_sizes,
+                                 const TensorTableEntry& entry);
+
+  // Store the Request for a name, and return whether the total count of
+  // Requests for that tensor is now equal to the HOROVOD size (and thus we are
+  // ready to reduce the tensor).
+  bool IncrementTensorCount(const Request& msg);
+
+  int rank_ = 0;
+  int local_rank_ = 0;
+  int cross_rank_ = 0;
+  int size_ = 1;
+  int local_size_ = 1;
+  int cross_size_ = 1;
+  bool is_coordinator_ = false;
+  bool is_homogeneous_ = false;
+
+  // ranks of the horovod world
+  std::vector<int> ranks_;
+
+  // COMM_WORLD ranks of processes running on this node.
+  std::vector<int> local_comm_ranks_;
+
+  // Numbers of ranks running per node
+  std::vector<int> local_sizes_for_cross_rank_;
+
+  uint32_t cache_capacity_ = 1024;
+
+  StallInspector stall_inspector_;
+
+  // Only exists on the coordinator node (rank zero). Maintains a count of
+  // how many nodes are ready to allreduce every tensor (keyed by tensor
+  // name) and time point when tensor started allreduce op.
+  MessageTable message_table_;
+
+  bool timeline_enabled_ = false;
+
+  // Outside dependencies
+  TensorQueue& tensor_queue_;
```

Comment:
One of the hotly debated topics in C++ is whether you should use references or pointers for member data (see [here](https://stackoverflow.com/questions/892133/should-i-prefer-pointers-or-references-in-member-data)).

There are a few notable differences between the two approaches that are worth considering:

- References cannot be re-assigned or null, unlike pointers.
- Objects containing references as members are non-assignable (cannot be copied).
- References cannot be lazy initialized.

Personally, I think it's better to use pointers, but it's your call.

Reply:
That's a good point. I think the reason I use references is that: 

- Those objects are actually owned by the global state, rather than shared by the controller and global state. So it would be misleading to use shared pointers.

- Because the lifetime of the global state is guaranteed to cover the lifetime of the controller. So by enforcing that those dependencies can only be assigned during initialization, we can ensure that the dependencies would always be valid. 

@alsrgv What do you think?